docs/Toy/toy/dialect.py

```diff
@@ -0,0 +1,267 @@
+                      func: Block.BlockCallback,
+                      private: bool = False):
+        type_attr = FunctionType.from_lists(input_types, return_types)
+        attributes = {
+            "sym_name": name,
+            "function_type": type_attr,
+        }
+        if private:
+            attributes["sym_visibility"] = "private"
+        return FuncOp.build(attributes=attributes,
+                            regions=[
+                                Region.from_block_list(
+                                    [Block.from_callable(input_types, func)])
+                            ])
+
+
+@irdl_op_definition
+class GenericCallOp(Operation):
+    name: str = "toy.generic_call"
+    arguments: Annotated[VarOperand, AnyAttr()]
+    callee: OpAttr[FlatSymbolRefAttr]
+
+    # Note: naming this results triggers an ArgumentError
+    res: Annotated[VarOpResult, AnyTensorTypeF64]
+
+    @classmethod
+    def get(cls: type[GenericCallOp], callee: Union[str, FlatSymbolRefAttr],
+            operands: List[Union[SSAValue, OpResult]],
+            return_types: List[Attribute]) -> GenericCallOp:
+        if isinstance(callee, str):
+            callee = FlatSymbolRefAttr.from_str(callee)
+
+        return super().create(operands=operands,
+                              result_types=return_types,
+                              attributes={"callee": callee})
+
+
+@irdl_op_definition
+class MulOp(Operation):
+    """
+    The "mul" operation performs element-wise multiplication between two
+    tensors. The shapes of the tensor operands are expected to match.
+    """
+    name: str = 'toy.mul'
+    arguments: Annotated[VarOperand, AnyTensorTypeF64]
+    res: Annotated[OpResult, AnyTensorTypeF64]
+
+    @classmethod
+    def from_summands(cls: type[MulOp], lhs: SSAValue, rhs: SSAValue) -> MulOp:
+        return super().create(result_types=[lhs.typ], operands=[lhs, rhs])
+
+    def verify_(self):
+        [arg0, arg1] = self.arguments
+        arg0_typ, arg1_typ = arg0.typ, arg1.typ
+        if isinstance(arg0_typ, TensorType) and isinstance(
+                arg1_typ, TensorType):
+            if arg0_typ.shape != arg1_typ.shape:
+                raise VerifyException(
+                    "Expected MulOp args to have the same shape")
+
+
+@irdl_op_definition
+class PrintOp(Operation):
+    """
+    The "print" builtin operation prints a given input tensor, and produces
+    no results.
+    """
+    name: str = 'toy.print'
+    arguments: Annotated[VarOperand, AnyAttr()]
+
+    @classmethod
+    def from_input(cls: type[PrintOp], input: SSAValue) -> PrintOp:
+        return super().create(operands=[input])
+
+
+@irdl_op_definition
+class ReturnOp(Operation):
+    """
+    The "return" operation represents a return operation within a function.
+    The operation takes an optional tensor operand and produces no results.
+    The operand type must match the signature of the function that contains
```

Comment:
Can you assert this?

Reply:
- [ ] verify returnop

Reply:
Should that be verified in the return op itself or by the containing function?

Reply:
Yep, added it in the funcop verifier